### PR TITLE
server: honor and validate the service mode for SQL pods

### LINKED ISSUE #83650: sql: connecting to an inactive tenant  should return an error
Context: https://cockroachlabs.slack.com/archives/C02HWA24541/p1656595977592689

For the purposes of streaming replication, we want to create a tenant in an inactive state such that during the entire replication process one is unable to query the state of the tenant.  This way the potentially-inconsistent state as we stream SSTables over is not observable until we explicitly revert to a single Cutover timestamp that leaves the tenant's tables in a consistent state and we call `sql.ActivateTenant` afterwards.

At the moment https://github.com/cockroachdb/cockroach/pull/83646 creates a tenant with the `descpb.TenantInfo_ADD` state and calls `ActivateTenant` afterwards, however

### LINKED ISSUE #93145: server: blocked mixed-style deployments [CRDB-14537 followup]
Requested by @andreimatei and @dhartunian 

For the time being, we are going to work under the assumption that for a given tenant, either

-  all SQL servers are in-process with KV, 
- OR all SQL servers are out-of-process.

but not both. 

As of v23.1, our main plan to enforce this assumption is via runbooks (i.e. tell users "don't do that").

However, we probably want a check in the product to prevent mistakes.

This could be done e.g. by looking at the sql instances/sessions table to see whether there's a session open.

Jira issue: CRDB-22187

Epic CRDB-26691

### PR DESCRIPTION
**Pull request by knz**
Rebased on top of #105441.
Fixes #93145.
Fixes #83650.

Epic: CRDB-26691

TLDR: this makes SQL servers only accept to start if their data state
is "ready" and their deployment type (e.g. separate process) matches
the configured service mode in the tenant record.
Additionally, the SQL server spontaneously terminates if
their service mode or data state is not valid any more (e.g
as a result of DROP TENANT or ALTER TENANT STOP SERVICE).

----

Prior to this patch, there wasn't a good story about the lifecycle
of separate-process SQL servers ("SQL pods"):

- if a SQL server was started against a non-existent tenant,
  an obscure error would be raised (`database "[1]" does not exist`)
- if a SQL server was started while a tenant was being added
  (keyspace not yet valid), no check would be performed and
  data corruption could ensue.
- if the tenant record/keyspace was dropped while the SQL server was
  running, SQL clients would start encountering obscure errors.

This commit fixes the situation by checking the tenant metadata:

- once, during SQL server startup, at which point server startup
  is prevented if the service check fails;
- then, asynchronously, whenever the metadata is updated, such that
  any service check failure results in a graceful shutdown of the SQL
  service.

The check proper validates:
- that the tenant record exists;
- that the data state is "ready";
- that the configured service mode matches that requested by the SQL
  server.

Examples output upon error:

- non-existent tenant:
  ```
  tenant service check failed: missing record
  ```

- attempting to start separate-process server while
  tenant is running as shared-process:
  ```
  tenant service check failed: service mode check failed: expected external, record says shared
  ```

- after ALTER TENANT STOP SERVICE:
  ```
  tenant service check failed: service mode check failed: expected external, record says none
  ```

- after DROP TENANT:
  ```
  tenant service check failed: service mode check failed: expected external, record says dropping
  ```

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
@stevendanna @arulajmani could you have a look at this PR before we invest in tests.

Currently it uses a unary TenantServiceCheck which is called periodically. I think this would be better served _eventually_ by a streaming RPC that hooks into a rangefeed over the tenant record. Do we want to do this right away? Should we go the way of this PR now, and go for the streaming RPC later?

For the rangefeed approach, I think I will need some help. I feel that this should be integrated with the capability watched somehow but I'm not sure about the APIs.

**Comment by stevendanna** [comment]
I think we'll probably need to rewrite some streaming tests for this. I can help with that this week.

**Comment by stevendanna** [comment]
>  I think this would be better served eventually by a streaming RPC that hooks into a rangefeed over the tenant record. Do we want to do this right away? Should we go the way of this PR now, and go for the streaming RPC later?

I'm OK with this simple approach for now. Mostly because I would like to see us think about the tenant process lifecycle a bit more before dedicating ourselves to a design.  I have questions such as whether service mode is the right abstraction in the long run. If so, do we need some system that avoids a situation in which a change in service mode doesn't allow for external and shared process sql servers running concurrently? Etc. These feel like longer run questions that we don't need to answer today to make some progress.

**Comment by knz** [comment]
RFAL

> do we need some system that avoids a situation in which a change in service mode doesn't allow for external and shared process sql servers running concurrently?

The new version of the PR answers this question: now that the service mode is pushed by KV to the SQL service, if we ever desire to support mixed deployments we can have KV servers push different values to different clients.

**Comment by knz** [comment]
@stevendanna @yuzefovich RFAL.

(Before this can merge, we will need to do something about the 3s startup overhead. In this PR, this overhead is now incurred for all uses of the test tenant, not just those that use nodelocal.)

**Review by stevendanna — COMMENTED** [review]
Overall this seems like a reasonable step forward to me.  

We should double check we are prepared to have this running in Serverless.

**Review by yuzefovich — COMMENTED** [review]
LGTM, but I'll defer to Steven for approval.

> Reviewed 1 of 5 files at r9, 46 of 46 files at r13, 41 of 41 files at r14, 41 of 41 files at r15, 45 of 45 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96144#-:-Na92iZP-DJiotDgUdVd:bv8hvps)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @nkodali, @renatolabs, @smg260, and @stevendanna)

___
*[`-- commits` line 32 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/96144#-Na90W8z1QLR0vcIta6P):*
nit: something is off in "only accept to start".
___
*[`pkg/server/server_sql.go` line 1702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/96144#gh-1272107446:-Na9193U-M8kjXO__IRz:b-rrayz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/069e1babde4bead2c101c97b19df110f3da5d435/pkg/server/server_sql.go#L1702)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Here and elsewhere could be an opportunity to adopt the new `must` library.
</blockquote></details>

I'm guessing that we'll want to backport this change (right?), and `must` library won't be available on 23.1 (it's not available on master yet either).

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 14 files at r1, 1 of 46 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nkodali, @renatolabs, @smg260, @stevendanna, and @yuzefovich)

___
*[`-- commits` line 32 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/96144#-Na90W8z1QLR0vcIta6P):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: something is off in "only accept to start".
</blockquote></details>

Adjusted.
___
*[`pkg/server/server_sql.go` line 1702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/96144#gh-1272107446:-NaBimTv1c7M89PicUjv:bq4xbfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/069e1babde4bead2c101c97b19df110f3da5d435/pkg/server/server_sql.go#L1702)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I'm guessing that we'll want to backport this change (right?), and `must` library won't be available on 23.1 (it's not available on master yet either).
</blockquote></details>

I'm ok with changing the assertion code during an eventual backport, but yahor is right and `must` hasn't merged yet.
___
*[`pkg/server/server_sql.go` line 1728 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/96144#gh-1272112064:-NaBjev66gGsK5M_ej1Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/069e1babde4bead2c101c97b19df110f3da5d435/pkg/server/server_sql.go#L1728)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Minor, but I could see moving this to a `s.checkServerModeMatchesEntry(entry)`
</blockquote></details>

Done.
___
*[`pkg/server/testserver.go` line 1086 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/96144#gh-1272115815:-NaBmHrD70lB_PiNRicv:b-a719yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/069e1babde4bead2c101c97b19df110f3da5d435/pkg/server/testserver.go#L1086)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we add a check for the service state into SkipTenantCheck below?
</blockquote></details>

It seems unnecessary, since the error you get if it's not right is pretty clear about what's going on.

But maybe you'd like to suggest, instead of a _check_, that `StartTenant` should proactively run `ALTER TENANT START SERVICE EXTERNAL`, perhaps unless a testing knob is set?

**Comment by knz** [comment]
> We should double check we are prepared to have this running in Serverless.

Internal conversation [here](https://cockroachlabs.slack.com/archives/C02HWA24541/p1690284395556829).

@stevendanna it seems to me we are in the clear. WDYT?

**Review by stevendanna — APPROVED** [review]

**Comment by knz** [comment]
TFYR

bors r=yuzefovich,stevendanna
